File: java/annalyns-infiltration/src/main/java/AnnalynsInfiltration.java
class AnnalynsInfiltration {
    public static boolean canFastAttack(boolean knightIsAwake) {
        return knightIsAwake ? false : true;
    }

    public static boolean canSpy(
            boolean knightIsAwake, boolean archerIsAwake, boolean prisonerIsAwake) {
        return knightIsAwake || archerIsAwake || prisonerIsAwake ? true : false;
    }

    public static boolean canSignalPrisoner(boolean archerIsAwake, boolean prisonerIsAwake) {
        return prisonerIsAwake && !archerIsAwake ? true : false;
    }

    public static boolean canFreePrisoner(
            boolean knightIsAwake,
            boolean archerIsAwake,
            boolean prisonerIsAwake,
            boolean petDogIsPresent) {
        return petDogIsPresent
                ? !archerIsAwake ? true : false
                : prisonerIsAwake && !knightIsAwake && !archerIsAwake ? true : false;
    }
}

File: julia/annalyns-infiltration/annalyns-infiltration.jl
function can_do_fast_attack(knight_awake)
    return !knight_awake
end

function can_spy(knight_awake, archer_awake, prisoner_awake)
    return knight_awake || archer_awake || prisoner_awake
end

function can_signal_prisoner(archer_awake, prisoner_awake)
    return prisoner_awake && !archer_awake
end

function can_free_prisoner(knight_awake, archer_awake, prisoner_awake, dog_present)
    return dog_present && !archer_awake || !dog_present && prisoner_awake &&!knight_awake && !archer_awake
end
